Sources/LiveKit/Core/Engine.swift

```diff
@@ -236,27 +265,49 @@ class Engine: MulticastDelegate<EngineDelegate> {
             }
         }
     }
+}
 
-    private func ensurePublisherConnected () -> Promise<Void> {
+// MARK: - Wait extension
 
-        guard let publisher = publisher else {
-            return Promise(EngineError.invalidState("publisher is nil"))
+extension Engine {
+
+    func waitForPublisherDataChannelOpen(reliability: Reliability, allowCurrentValue: Bool = true) -> Promise<Void> {
+
+        guard let dcPublisher = publisherDataChannel(for: reliability) else {
+            return Promise(EngineError.invalidState("publisher data channel is nil"))
         }
 
-        guard subscriberPrimary, !publisher.isIceConnected else {
-            // aleady connected, no-op
+        logger.debug("waiting for dataChannel to open on \(dcPublisher)")
+        if allowCurrentValue, dcPublisher.readyState == .open {
+            logger.debug("dataChannel already open")
             return Promise(())
         }
 
-        publisherShouldNegotiate()
-
-        return waitForIceConnect(transport: publisher)
+        return Promise<Void> { resolve, fail in
+            // create temporary delegate
+            var engineDelegate: EngineDelegateClosures?
+            engineDelegate = EngineDelegateClosures(
+                onDataChannelStateUpdated: { _, dataChannel, state in
+                    if dataChannel == dcPublisher, state == .open {
+                        resolve(())
+                        engineDelegate = nil
+                    }
+                }
+            )
+            self.add(delegate: engineDelegate!)
+            // detect signal close while waiting
+            var signalDelegate: SignalClientDelegateClosures?
+            signalDelegate = SignalClientDelegateClosures(
+                didClose: { _, _ in
+                    fail(SignalClientError.close("Socket closed while waiting for ice-connect"))
+                    signalDelegate = nil
+                }
+            )
+            self.signalClient.add(delegate: signalDelegate!)
```

Comment:
does this get removed automatically?

Reply:
This should get cleaned up automatically since it's only weak reference.